### DISCUSSION TITLE
Is Airflow the right tool for a chronjob that runs every minute or so (and needs to scale to 100s of concurrent tasks)?

### DISCUSSION THREAD

**Discussion by p-hebert** [apache/airflow#26558] [Q&A] [answered]
Hello!

We are investigating the addition of a poor-man's change data capture against a set of remote databases. Due to several constraints (limited permissions, legacy database), our only solution available is polling of the table for updated/created records.
We need a near-real time read, but once we have the data we can process it as batch once a day without issue.

We would need to poll each of these remote databases once every 1-3 minutes between 7AM and 9PM, and we have ~50 such remote databases, number which will increase to 100, then 100s over the next year.
We are using Airflow with Celery scheduler on K8S.

With the above information, do you think Airflow would be the right tool for this job, or would we do better to schedule cloud functions on a chrontab?

**Comment by potiuk** [comment, marked as answer]
It should work, not ideal but it will. But It will use a lot of resources (but having a crontab running cloud functions will be likely not far-off from that if you run them that frequently - I think and far less manageable). 

I tihnk the best approach for you will be the combination of sensors - possibly with poke_reschedule + dags with max_active_runs=1 and catchup=False, should do the job. https://stackoverflow.com/questions/71502271/how-to-skip-next-schedule-of-dag-if-previous-run-not-yet-finished-by-the-time-of   (You can look all of it up in Airflow docs).

Not sure about the cost comparision - this you have to calculate yourself.

Also Airlfow will be good if you have some more or less complex logic to follow 

Ideally - if you would be able to make your queries utilise Async IO rather than active polling, it would save you enormous amounf of resources to run those sensors - by using Triggerer and Deferreable Operators https://airflow.apache.org/docs/apache-airflow/stable/concepts/deferring.html - but implementing your operators in the way that might turn your operators into Defrerrables in the future you might have an easy migration path in the future if your constraints are lifted. 

One solution for such future constraints lifting is to extend your legacy database and add some async-io capable proxy that would be able to poll the DB locally and expose an async.io - capable interface.

**Reply by p-hebert** [reply]
Hey @potiuk !

Thanks for your thorough answer & sorry for the delay in response.

RE: Not ideal: By not ideal, do you refer to the amount of resources used to perform such an operation?

RE: Cloud Function alternative - I think we misunderstood each other regarding the alternative solution of using Cloud Function - I'd keep this entirely separate from Airflow and leverage the native support for Cloud Function invocation via crontab. I'd probably integrate some sort of monitoring DAG every 15 min to ensure we don't have breakage.

RE: AsyncIO - that's a great tip. I'll check if we can make our Sensors async - we use google-cloud-pubsub to interact with our extraction agent located on the servers where the dbs reside.

For the time being I think we'll opt for the Airflow implementation and see if it scales nicely and otherwise move to Cloud Functions.
